docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+`FontData` object (a wrapper around a rust byte slice (`&[u8]`), essentially
+a pointer plus a length). It provides typed access to that tables fields.
+
+Conceptually, this looks like:
+
+```rust
+pub struct MyTable<'a>(FontData<'a>);
+
+impl MyTable<'_> {
+    /// Read the table's first field
+    pub fn format(&self) -> u16 {
+        self.0.read_at(0)
+    }
+}
+```
+
+In practice, what we generate is slightly different: instead of
+generating a struct for the table itself (and wrapping the data directly)
+we generate a 'marker' struct, which defines the type of the table, and then we
+combine it with the data via a `TableRef` struct.
+
+The `TableRef` struct looks like this:
+
+```rust
+/// Typed access to raw table data.
+pub struct TableRef<'a, T> {
+    shape: T,
+    data: FontData<'a>,
+}
+```
+
+And the definition of the table above, using a marker type, would look something
+like:
+
+```rust
+/// A marker type
+pub struct MyTableMarker;
+
+/// Instead of generating a struct for each table, we define a type alias
+pub type MyTable<'a> = TableRef<'a, MyTableMarker>;
+
+impl MyTableMarker {
+    fn format_byte_range(&self) -> Range<usize> {
+        0..u16::RAW_BYTE_LEN
+    }
+}
+
+impl MyTable<'_> {
+    fn format(&self) -> u16 {
+        let range = self.shape.format_byte_range();
+        self.data.read_at(range.start)
+    }
+}
+```
+
+To the user these two API are equivalent (you have a type `MyTable`, on which
+you can call methods to read fields) but the 'marker' pattern potentially allows
+for us to do some fancy things in the future (involving various cases where we
+want to store a type separate from a lifetime).
+
+> ***note:***
+>
+> there are also downsides of the marker pattern; in particular, currently
+> the code we generate will only compile if it is part of the `read-fonts` crate
+> itself. This isn't a major limitation, except that it makes certain kinds of
+> testing harder to do, since we can't do fancy things like generate code that
+> treated as a separate compilation unit, e.g. for use with the [`trybuild`][]
+crate.
+
+#### <a id="font-read-args"></a> `FontRead` & `FontReadWithArgs`
+
+After generating the type definitions, the next thing we generate is an
+implementation of one of [`FontRead`][] or [`FontReadWithArgs`][]. The
+`FontRead` trait is used if a table is self-describing: that is, if the data in
+the table can be fully interpreted without any external information. In some
+cases, however, this is not possible. A simple example is the [`loca` table][loca-spec]:
+the data for this table cannot be interpreted correctly without knowing the
+number of glyphs in the font (stored in the `maxp` table) as well as whether the
+format is long or short, which is stored in the `head` table.
+
+In either case, the generated table code is very similar.
+
+For the purpose of illustration, let's imagine we have a table that looks like
+this:
+
+```rust
+table Foob {
```

Comment:
I suggest doing a real example might be helpful for a reader familiar with fonts and less familiar with Rust. I think many readers will fit that description. OS/2 with some fields ommitted perhaps?